File: bifurcate/container.py
"""The .bfc container format and the end-to-end API.

Layout (every field authenticated by the trailing tag):

    offset  field
    0       magic        "BFC1"          4
    4       version      0x01            1
    5       flags        mode | hybrid   1
    6       salt                        16
    22      iv                          16
    38      wrapped_len  uint16 BE       2
    40      wrapped_key  (0 or 256)     var
    ..      ciphertext                  var
    end-32  hmac_tag                    32

Self-describing files are the determinism proof: you can encrypt on one
machine and decrypt on another with no out-of-band parameters.
"""

from __future__ import annotations

import hashlib
import os
from dataclasses import dataclass

from .kdf import derive_subkeys
from .keyschedule import expand
from .mac import TAG_BYTES, IntegrityError, compute_tag, verify_tag
from .modes import DECRYPTORS, ENCRYPTORS, MODE_CBC, MODE_ECB, MODE_NAMES

MAGIC = b"BFC1"
VERSION = 1
SALT_BYTES = 16
IV_BYTES = 16
HEADER_LEN = 40  # magic..wrapped_len inclusive

FLAG_ECB = 0b0000_0001
FLAG_HYBRID = 0b0000_0010

PBKDF2_ITERATIONS = 200_000


class FormatError(Exception):
    """Raised when a file is not a valid .bfc container."""


# -- key material -----------------------------------------------------------
def root_key_from_passphrase(passphrase: str, salt: bytes) -> bytes:
    """PBKDF2-HMAC-SHA256, 200k iterations.

    Stretches a low-entropy human passphrase to a full 256-bit key and makes
    dictionary attacks expensive. The chaos engine is NEVER seeded from a raw
    passphrase.
    """
    return hashlib.pbkdf2_hmac(
        "sha256", passphrase.encode("utf-8"), salt, PBKDF2_ITERATIONS, dklen=32
    )


@dataclass(frozen=True)
class Container:
    """A parsed .bfc file."""

    version: int
    flags: int
    salt: bytes
    iv: bytes
    wrapped_key: bytes
    ciphertext: bytes
    tag: bytes

    @property
    def mode(self) -> int:
        return MODE_ECB if self.flags & FLAG_ECB else MODE_CBC

    @property
    def mode_name(self) -> str:
        return MODE_NAMES[self.mode]

    @property
    def is_hybrid(self) -> bool:
        return bool(self.flags & FLAG_HYBRID)


def _pack(flags, salt, iv, wrapped_key, ciphertext) -> bytes:
    """Serialise everything except the tag (this is exactly the MAC input)."""
    return (
        MAGIC
        + bytes([VERSION, flags])
        + salt
        + iv
        + len(wrapped_key).to_bytes(2, "big")
        + wrapped_key
        + ciphertext
    )


# -- encrypt ----------------------------------------------------------------
def encrypt(
    plaintext: bytes,
    *,
    passphrase: str | None = None,
    public_key=None,
    mode: int = MODE_CBC,
) -> bytes:
    """Encrypt `plaintext` into a complete .bfc container.

    Exactly one of `passphrase` or `public_key` must be given.
    """
    if (passphrase is None) == (public_key is None):
        raise ValueError("provide exactly one of passphrase or public_key")

    salt = os.urandom(SALT_BYTES)
    iv = os.urandom(IV_BYTES)
    flags = 0
    if mode == MODE_ECB:
        flags |= FLAG_ECB

    if public_key is not None:
        from .hybrid import new_root_key, wrap_key

        root_key = new_root_key()
        wrapped = wrap_key(public_key, root_key)
        flags |= FLAG_HYBRID
    else:
        root_key = root_key_from_passphrase(passphrase, salt)
        wrapped = b""

    enc_seed, mac_key = derive_subkeys(root_key)
    rk = expand(enc_seed)
    ciphertext = ENCRYPTORS[mode](plaintext, rk, iv)

    body = _pack(flags, salt, iv, wrapped, ciphertext)
    return body + compute_tag(mac_key, body)


# -- parse / decrypt --------------------------------------------------------
def parse(blob: bytes) -> Container:
    """Parse a .bfc file without verifying or decrypting it."""
    if len(blob) < HEADER_LEN + TAG_BYTES:
        raise FormatError("file too short to be a .bfc container")
    if blob[:4] != MAGIC:
        raise FormatError("bad magic — not a .bfc container")
    version = blob[4]
    if version != VERSION:
        raise FormatError(f"unsupported container version {version}")
    flags = blob[5]
    salt = blob[6:22]
    iv = blob[22:38]
    wrapped_len = int.from_bytes(blob[38:40], "big")
    start = HEADER_LEN
    wrapped_key = blob[start : start + wrapped_len]
    if len(wrapped_key) != wrapped_len:
        raise FormatError("truncated wrapped key")
    ciphertext = blob[start + wrapped_len : -TAG_BYTES]
    tag = blob[-TAG_BYTES:]
    return Container(version, flags, salt, iv, wrapped_key, ciphertext, tag)


def decrypt(
    blob: bytes,
    *,
    passphrase: str | None = None,
    private_key=None,
) -> bytes:
    """Verify then decrypt a .bfc container.

    The tag is checked BEFORE decryption. On failure this raises
    :class:`IntegrityError` — it never returns garbage.
    """
    c = parse(blob)

    if c.is_hybrid:
        if private_key is None:
            raise ValueError("this container is hybrid — an RSA private key is required")
        from .hybrid import unwrap_key

        try:
            root_key = unwrap_key(private_key, c.wrapped_key)
        except Exception as exc:  # wrong private key
            raise IntegrityError("RSA unwrap failed — wrong private key") from exc
    else:
        if passphrase is None:
            raise ValueError("this container needs a passphrase")
        root_key = root_key_from_passphrase(passphrase, c.salt)

    enc_seed, mac_key = derive_subkeys(root_key)

    # Authenticate the entire prefix before touching the ciphertext.
    verify_tag(mac_key, blob[:-TAG_BYTES], c.tag)

    rk = expand(enc_seed)
    return DECRYPTORS[c.mode](c.ciphertext, rk, c.iv)

File: cipher-ts/src/container.ts
// The .bfc container — bifurcate/container.py. Byte-compatible layout.
import { deriveSubkeys, rootKeyFromPassphrase } from "./kdf.js";
import { expand } from "./keyschedule.js";
import { cbcEncrypt, cbcDecrypt, ecbEncrypt, ecbDecrypt, MODE_CBC, MODE_ECB } from "./modes.js";
import { computeTag, verifyTag, TAG_BYTES, IntegrityError } from "./mac.js";
import { concat } from "./bytes.js";

const MAGIC = new Uint8Array([0x42, 0x46, 0x43, 0x31]); // "BFC1"
const VERSION = 1;
const HEADER_LEN = 40;
export const FLAG_ECB = 0b01;
export const FLAG_HYBRID = 0b10;
export const PBKDF2_ITERATIONS = 200_000;

export class FormatError extends Error {}
export { IntegrityError };

export interface Parsed {
  version: number; flags: number; salt: Uint8Array; iv: Uint8Array;
  wrappedKey: Uint8Array; ciphertext: Uint8Array; tag: Uint8Array;
  mode: number; modeName: string; isHybrid: boolean;
}

const rand = (n: number): Uint8Array => crypto.getRandomValues(new Uint8Array(n));

function pack(flags: number, salt: Uint8Array, iv: Uint8Array, wrapped: Uint8Array, ct: Uint8Array): Uint8Array {
  const wl = new Uint8Array([(wrapped.length >> 8) & 0xff, wrapped.length & 0xff]);
  return concat(MAGIC, new Uint8Array([VERSION, flags]), salt, iv, wl, wrapped, ct);
}

export interface EncryptOpts { passphrase?: string; publicKey?: CryptoKey; mode?: number; }

export async function encrypt(plaintext: Uint8Array, opts: EncryptOpts): Promise<Uint8Array> {
  const mode = opts.mode ?? MODE_CBC;
  if (!!opts.passphrase === !!opts.publicKey) throw new Error("provide exactly one of passphrase or publicKey");
  const salt = rand(16), iv = rand(16);
  let flags = mode === MODE_ECB ? FLAG_ECB : 0;
  let rootKey: Uint8Array, wrapped = new Uint8Array(0);
  if (opts.publicKey) {
    rootKey = rand(32);
    wrapped = new Uint8Array(await crypto.subtle.encrypt({ name: "RSA-OAEP" }, opts.publicKey, rootKey));
    flags |= FLAG_HYBRID;
  } else {
    rootKey = rootKeyFromPassphrase(opts.passphrase!, salt, PBKDF2_ITERATIONS);
  }
  const { encSeed, macKey } = deriveSubkeys(rootKey);
  const rk = expand(encSeed);
  const ct = mode === MODE_ECB ? ecbEncrypt(plaintext, rk) : cbcEncrypt(plaintext, rk, iv);
  const body = pack(flags, salt, iv, wrapped, ct);
  return concat(body, computeTag(macKey, body));
}

export function parse(blob: Uint8Array): Parsed {
  if (blob.length < HEADER_LEN + TAG_BYTES) throw new FormatError("file too short");
  for (let i = 0; i < 4; i++) if (blob[i] !== MAGIC[i]) throw new FormatError("bad magic — not a .bfc container");
  const version = blob[4]!;
  if (version !== VERSION) throw new FormatError(`unsupported version ${version}`);
  const flags = blob[5]!;
  const salt = blob.slice(6, 22), iv = blob.slice(22, 38);
  const wl = (blob[38]! << 8) | blob[39]!;
  const wrappedKey = blob.slice(HEADER_LEN, HEADER_LEN + wl);
  if (wrappedKey.length !== wl) throw new FormatError("truncated wrapped key");
  const ciphertext = blob.slice(HEADER_LEN + wl, blob.length - TAG_BYTES);
  const tag = blob.slice(blob.length - TAG_BYTES);
  const isHybrid = !!(flags & FLAG_HYBRID);
  const mode = flags & FLAG_ECB ? MODE_ECB : MODE_CBC;
  return { version, flags, salt, iv, wrappedKey, ciphertext, tag, mode,
           modeName: mode === MODE_ECB ? "ECB" : "CBC", isHybrid };
}

export interface DecryptOpts { passphrase?: string; privateKey?: CryptoKey; }

export async function decrypt(blob: Uint8Array, opts: DecryptOpts): Promise<Uint8Array> {
  const c = parse(blob);
  let rootKey: Uint8Array;
  if (c.isHybrid) {
    if (!opts.privateKey) throw new Error("this container is hybrid — an RSA private key is required");
    try {
      rootKey = new Uint8Array(await crypto.subtle.decrypt({ name: "RSA-OAEP" }, opts.privateKey, c.wrappedKey));
    } catch (e) { throw new IntegrityError("RSA unwrap failed — wrong private key"); }
  } else {
    if (opts.passphrase == null) throw new Error("this container needs a passphrase");
    rootKey = rootKeyFromPassphrase(opts.passphrase, c.salt, PBKDF2_ITERATIONS);
  }
  const { encSeed, macKey } = deriveSubkeys(rootKey);
  verifyTag(macKey, blob.slice(0, blob.length - TAG_BYTES), c.tag); // authenticate BEFORE decrypting
  const rk = expand(encSeed);
  return c.mode === MODE_ECB ? ecbDecrypt(c.ciphertext, rk) : cbcDecrypt(c.ciphertext, rk, c.iv);
}
